zebra-test/src/transcript.rs

```diff
@@ -1,59 +1,144 @@
 where
-    I: Iterator<Item = (R, S)>,
+    I: Iterator<Item = (R, Result<S, E>)>,
     R: Debug,
     S: Debug + Eq,
+    E: Fn(Error) -> Result<(), Error>,
 {
     pub async fn check<C>(mut self, mut to_check: C) -> Result<(), Report>
     where
         C: Service<R, Response = S>,
-        C::Error: Debug,
+        C::Error: Into<Error>,
     {
         while let Some((req, expected_rsp)) = self.messages.next() {
             // These unwraps could propagate errors with the correct
             // bound on C::Error
-            let rsp = to_check.ready_and().await.unwrap().call(req).await.unwrap();
-            ensure!(
-                rsp == expected_rsp,
-                "Expected {:?}, got {:?}",
-                expected_rsp,
-                rsp
-            );
+            let fut = to_check.ready_and().await;
+
+            let fut = match fut {
+                Ok(fut) => Ok(fut),
+                Err(e) => match &expected_rsp {
+                    Ok(expected_rsp) => Err(eyre!(e.into()))
+                        .wrap_err("service's ready_and returned an error: transcript expected an response")
+                        .with_section(|| {
+                            format!("{:?}", expected_rsp).header("Expected Response:")
+                        }),
+                    Err(error_checker) => {
+                        error_checker(e.into())
+                            .map_err(|e| eyre!(e))
+                            .wrap_err("service returned an error but it didn't match the expected error")?;
+                        continue;
+                    },
+                },
+            }?;
+
+            let response = fut.call(req).await;
+
+            match (response, expected_rsp) {
+                (Ok(rsp), Ok(expected_rsp)) => {
+                    if rsp != expected_rsp {
+                        Err(eyre!(
+                            "response doesn't match transcript's expected response"
+                        ))
+                        .with_section(|| format!("{:?}", expected_rsp).header("Expected Response:"))
+                        .with_section(|| format!("{:?}", rsp).header("Found Response:"))?;
+                    }
+                }
+                (Ok(rsp), Err(_)) => {
+                    todo!("got response when an error was expected. rsp={:?}", rsp)
+                }
+                (Err(_), Ok(expected_rsp)) => todo!(
+                    "got an error when a response was expected. expected_rsp={:?}",
+                    expected_rsp
+                ),
+                (Err(e), Err(error_checker)) => {
+                    error_checker(e.into()).map_err(|e| eyre!(e)).wrap_err(
+                        "service returned an error but it didn't match the expected error",
+                    )?;
+                    continue;
+                }
+            }
         }
         Ok(())
     }
 }
 
-impl<R, S, I> Service<R> for Transcript<R, S, I>
+const TRANSCRIPT_MOCK_GUIDE: &str = r#"
```

Comment:
This is what I came up with for handling errors when using transcripts as a mock. LMK what you think @hdevalence

Reply:
I don't think I understand why the guide is a static str rather than part of the `Transcript` documentation.

Reply:
Because this is runtime information that is output as a custom section in the error report when someone writes a test that is clearly bugged.